### PR TITLE
[lldb] Add QSupported key to report watchpoint types supported

### PR DESCRIPTION
**Pull request by jasonmolenda**
debugserver on arm64 devices can manage both Byte Address Select watchpoints (1-8 bytes) and MASK watchpoints (8 bytes-2 gigabytes). This adds a SupportedWatchpointTypes key to the QSupported response from debugserver with a list of these, so lldb can take full advantage of them when creating larger regions with a single hardware watchpoint.

Also add documentation for this, and two other lldb extensions, to the lldb-gdb-remote.txt documentation.

Re-enable TestLargeWatchpoint.py on Darwin systems when testing with the in-tree built debugserver.  I can remove the "in-tree built debugserver" in the future when this new key is handled by an Xcode debugserver.

### CODE REVIEW

**Comment by jasonmolenda** [comment]
The QSupported packet from debugserver also has a "compression minimum size" field it adds (see the change to RNBRemote.cpp) and the QEnableCompression packet accepts that field as well.  I must have had a reason for adding this when I added compression to the protocol, but lldb doesn't use this.  I'll remove it in a separate patch.

File: lldb/docs/lldb-gdb-remote.txt
Comment (DavidSpickett) [inline lldb/docs/lldb-gdb-remote.txt:59]: Is the total number of lldb extensions 3? Might as well say "the example shows all the extensions".

If you don't know (I don't either) how many extensions we have, let's imagine I never left this comment :)

File: lldb/docs/lldb-gdb-remote.txt
Comment (DavidSpickett) [inline lldb/docs/lldb-gdb-remote.txt:57]: Is there really a space in `SupportedWatchpointTypes =aarch64-mask` or just a typo?

File: lldb/docs/lldb-gdb-remote.txt
Comment (DavidSpickett) [inline lldb/docs/lldb-gdb-remote.txt:61]: "base 16"?

Also what is a serial stub? Does that literally mean when we use a serial port to talk to the debug server?

File: lldb/docs/lldb-gdb-remote.txt
Comment (DavidSpickett) [inline lldb/docs/lldb-gdb-remote.txt:78]: I would rephrase and put "If nothing is specified" first, so the reader knows immediately what the point of this is.

Also tab it in to be part of the `SupportedWatchpointTypes` block.

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp
Comment (DavidSpickett) [inline lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:421]: Please put this string -> enum code next to the enum if possible. Perhaps `eWatchpointHardwareFeatureUnknown` could be the fallback value, assuming that that is 0.

Then you can loop over the strings here doing:
m_watchpoint_types |= stringToType(..);

And we'll have one place to update/reference in future.

Comment (DavidSpickett) [reply lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:421]: Also I wonder if this should be case insensitive, but I doubt it'll ever be an issue. If compression types does that then sure, if not, it's fine as is.

**Comment by DavidSpickett** [comment]
Just to record that I thought about it, I agree with not trying to make the names architecture neutral.

For example one might say that because range watchpoints on mips and aarch64 have overlapping functionality (no pun intended) that we could report that we have "range" as a type for either of them. This leads to a situation where the functionality that doesn't overlap becomes much harder to handle and detect.

So I agree with `aarch64-mask` instead of simply `mask`, as an example.

If we later want to (and kinda already do) group these by general functionality, we can do that in lldb in the algorithms you added for working out the watchpoint resources.

(and sod's law, someone will propose to add `aarch64-bas-but-this-one-thing-is-different` to the architecture one of these days :) )

File: lldb/docs/lldb-gdb-remote.txt
Comment (jasonmolenda) [reply lldb/docs/lldb-gdb-remote.txt:57]: lol, copy & paste had spread it across two lines.  I joined them but missed the space that got introduced.

File: lldb/docs/lldb-gdb-remote.txt
Comment (jasonmolenda) [reply lldb/docs/lldb-gdb-remote.txt:61]: I could say "the hex maximum packet size", I suppose, is that clearer?  One of the annoying things (IMO) about gdb remote serial protocol is that many numbers are hex numbers without a 0x prefix, but sometimes (probably mistakes) they're base 10.  So if you see a number like "2000" in a log file, you don't know what that number is.

The naming in this text file is a bit all over the place.  The protocol is called "gdb remote serial protocol".  The opposite side from lldb, debugserver or lldb-server, is called the stub.  Some places call it the "gdb RSP stub", some just say "stub".

Comment (jasonmolenda) [reply lldb/docs/lldb-gdb-remote.txt:61]: (incidentally debugserver also reports a compression minimum size -- which I'll be removing after this patch is landed because it does nothing -- and that number is sent as a base 10 number.  lol it's a wild west for numbers.)

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp
Comment (jasonmolenda) [reply lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:421]: I thought about putting the gdb remote serial protocol names in lldb-enumerations.h but they're gdb RSP specific strings and I don't expect them to be used anywhere else.

File: lldb/docs/lldb-gdb-remote.txt
Comment (jasonmolenda) [reply lldb/docs/lldb-gdb-remote.txt:59]: I don't have high confidence that the documentation will be updated as people add new extensions, so I was reluctant to make it sound like the extensions were exhaustive.

**Comment by jasonmolenda** [comment]
> Just to record that I thought about it, I agree with not trying to make the names architecture neutral.
> 
> For example one might say that because range watchpoints on mips and aarch64 have overlapping functionality (no pun intended) that we could report that we have "range" as a type for either of them. This leads to a situation where the functionality that doesn't overlap becomes much harder to handle and detect.
> 
> So I agree with `aarch64-mask` instead of simply `mask`, as an example.
> 
> If we later want to (and kinda already do) group these by general functionality, we can do that in lldb in the algorithms you added for working out the watchpoint resources.
> 
> (and sod's law, someone will propose to add `aarch64-bas-but-this-one-thing-is-different` to the architecture one of these days :) )

Yeah I thought about specifying a more structured name, like "ARCH-STYLE" so we'd have "x86_64-mask" in addition to "aarch64-mask", but my guess is a lot of targets will work fine with "power of 2, up to sizeof(void*)" and never need to adjust this.  Ted Woodward pointed out that the PowerPC watchpoints are quite different, so if we support that target we'll definitely need an additional algorithm for sure.

**Comment by jasonmolenda** [comment]
@bulbazord in the most recent commit I moved this internal-only enum from lldb-enumerations.h to lldb-private-enumerations.h, but I need to use the constexpr templatey thing that LLDB_MARK_AS_BITMASK_ENUM() defines for the enum so I can use bit-wise | & operations without casting everywhere; that's defined in lldb-enumerations.h so I included the public enums in the lldb-private-enumerations.h.  It seems like it's probably not a great choice, but the other one is breaking out this and FLAGS_ENUM etc into a little lldb-common-enumerations.h or something.  What do you think?

File: lldb/docs/lldb-gdb-remote.txt
Comment (DavidSpickett) [inline lldb/docs/lldb-gdb-remote.txt:62]: I was mostly confused by the lack of a space between base and 16. "base 16" is recognisable, "base16" sounds like a nightclub.I was mostly confused by the lack of a space between base and 16. "base 16" is recognisable, "base16" sounds like a nightclub.

File: lldb/docs/lldb-gdb-remote.txt
Comment (DavidSpickett) [inline lldb/docs/lldb-gdb-remote.txt:59]: "In the example above"

File: lldb/docs/lldb-gdb-remote.txt
Comment (DavidSpickett) [inline lldb/docs/lldb-gdb-remote.txt:52]: "can parse. They are not all listed here.
<blank line>
An example exchange:"

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp
Comment (DavidSpickett) [reply lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:417]: Sure. The other motivation for me was to not add more else if to `GetRemoteQSupported` which is probably a gazillion lines long. But a few more on top of that isn't changing anything so it's fine.

**Comment by bulbazord** [comment]
> @bulbazord in the most recent commit I moved this internal-only enum from lldb-enumerations.h to lldb-private-enumerations.h, but I need to use the constexpr templatey thing that LLDB_MARK_AS_BITMASK_ENUM() defines for the enum so I can use bit-wise | & operations without casting everywhere; that's defined in lldb-enumerations.h so I included the public enums in the lldb-private-enumerations.h. It seems like it's probably not a great choice, but the other one is breaking out this and FLAGS_ENUM etc into a little lldb-common-enumerations.h or something. What do you think?

Not the greatest thing in the world but it's not terrible either. We use the lldb public enumerations everywhere in private code, so as long as we're not going the other way this is ok to do. I will say, by moving the definitions from `lldb-enumerations.h` to `lldb-private-enumerations.h` you have removed these values from the python bindings. Technically that's an API break, but I'm not sure where anyone could have used these values otherwise. LGTM.

File: lldb/docs/lldb-gdb-remote.txt
Comment (jasonmolenda) [reply lldb/docs/lldb-gdb-remote.txt]: OK, I can see that.  I fixed this and a handful of other basenn uses in the file (probably all from me over the years).

**Comment by jasonmolenda** [comment]
Thanks for the feedback Alex & David, I think this one might be good enough now.

**Review by JDevlieghere — APPROVED** [review]
LGTM if @DavidSpickett and @bulbazord are happy.

**Review by bulbazord — APPROVED** [review]
LGTM

**Pull request merged**
